fs/xattr.c

```diff
@@ -653,6 +655,16 @@ static int path_setxattr(const char __user *pathname,
 	error = user_path_at(AT_FDCWD, pathname, lookup_flags, &path);
 	if (error)
 		goto out;
+#ifdef CONFIG_TRUENAS
+	if (size) {
+		if (size > XATTR_SIZE_MAX) {
+			if ((size > XATTR_LARGE_SIZE_MAX) ||
+			    (IS_LARGE_XATTR(path.dentry->d_inode) == 0)) {
+				return -E2BIG;
```

Comment:
I have feeling here are missing `path_put(&path)` and `kvfree(ctx.kvalue)`.

Reply:
While there, I wonder what was the point to check for `size` and then for `size > XATTR_SIZE_MAX`.  Though it is not new and insignificant.

Reply:
This looks like it got moved from where it should be because of refactoring in xattr.c by upstream. It should be in `do_setxattr`.

Reply:
When I was working on an updated 6.6 for testing purposes I did something like this:

```
commit 8a1aae17e6f54be3fda82a76a03f07004d7a730c
Author: Andrew Walker <awalker@ixsystems.com>
Date:   Thu Oct 31 16:24:00 2024 -0400

    Fix do_setxattr

diff --git a/fs/xattr.c b/fs/xattr.c
index ce0d11556bea..30f2a5822d4f 100644
--- a/fs/xattr.c
+++ b/fs/xattr.c
@@ -669,6 +669,13 @@ int do_setxattr(struct mnt_idmap *idmap, struct dentry *dentry,
                return do_set_acl(idmap, dentry, ctx->kname->name,
                                  ctx->kvalue, ctx->size);

+#ifdef CONFIG_TRUENAS
+       if (ctx->size &&
+           (ctx->size > XATTR_SIZE_MAX) &&
+           (IS_LARGE_XATTR(d->d_inode) == 0)) {
+               return -E2BIG;
+       }
+#endif
        return vfs_setxattr(idmap, dentry, ctx->kname->name,
                        ctx->kvalue, ctx->size, ctx->flags);
 }
@@ -677,17 +684,6 @@ static int path_setxattr(const char __user *pathname,
                         const char __user *name, const void __user *value,
                         size_t size, int flags, unsigned int lookup_flags)
 {
-#ifdef CONFIG_TRUENAS
-       if (size) {
-               if (size > XATTR_SIZE_MAX) {
-                       if ((size > XATTR_LARGE_SIZE_MAX) ||
-                           (IS_LARGE_XATTR(d->d_inode) == 0)) {
-                               return -E2BIG;
-                       }
-               }
-       }
-#endif
-
        struct xattr_name kname;
        struct xattr_ctx ctx = {
                .cvalue   = value,
                
```

Reply:
I have updated as you suggested, can you please take a look?

Reply:
As I said above, `ctx->size && (ctx->size > XATTR_SIZE_MAX)` is equivalent to just `ctx->size > XATTR_SIZE_MAX`.  And the indentation of the following line seems wrong.

Also, @anodos325, it seems in your proposed patch you've remove hard limit of `XATTR_LARGE_SIZE_MAX` in case of `IS_LARGE_XATTR(d->d_inode) != 0`. Was it intentional?

Reply:
I have updated removed the `ctx->size` check. Indentation should also be fixed. @anodos325 can comment about removing the hard limit of `XATTR_LARGE_SIZE_MAX`.

Reply:
We need hard limit of `XATTR_LARGE_SIZE_MAX`. Concur it can be simplified, that is artifact IIRC from back in days of angelfish (matching previous check for XATTR_SIZE_MAX).

Basically:
1. if inode supports large xattrs then we should check against XATTR_LARGE_SIZE_MAX
2. if inode does not support large xattrs then we should check XATTR_SIZE_MAX

Reply:
Looking at this again (been a few years), we may also want to set upper bound in `setxattr_copy` (unrelated to current questions):
```
diff --git a/fs/xattr.c b/fs/xattr.c
index c5a8a7c3fb4a..a9b12be7283f 100644
--- a/fs/xattr.c
+++ b/fs/xattr.c
@@ -646,7 +646,10 @@ int setxattr_copy(const char __user *name, struct xattr_ctx *ctx)

        error = 0;
        if (ctx->size) {
-#ifndef CONFIG_TRUENAS
+#ifdef CONFIG_TRUENAS
+               if (ctx->size > XATTR_LARGE_SIZE_MAX)
+                       return -E2BIG;
+#else
                if (ctx->size > XATTR_SIZE_MAX)
                        return -E2BIG;
 #endif
 ```

To avoid large `memdup_user()`. Thoughts?

Reply:
If we do above change then we can simplify check in `do_setxattr` to return -E2BIG if does not support large xattrs and xattr size is larger than XATTR_SIZE_MAX.

Reply:
Thanks, I have updated accordingly to check for hard limit of `XATTR_LARGE_SIZE_MAX` in `setxattr_copy`.

Reply:
The indentation is not right again.

Reply:
Sorry about that, it should be fixed now.